plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts

```diff
@@ -15,31 +15,79 @@
  */
 
 import {
+  EDIT_URL_ANNOTATION,
   Entity,
-  LOCATION_ANNOTATION,
   LocationSpec,
+  LOCATION_ANNOTATION,
   ORIGIN_LOCATION_ANNOTATION,
+  SOURCE_LOCATION_ANNOTATION,
+  VIEW_URL_ANNOTATION,
 } from '@backstage/catalog-model';
+import { ScmIntegrationRegistry } from '@backstage/integration';
 import lodash from 'lodash';
 import { CatalogProcessor, CatalogProcessorEmit } from './types';
+import gitUrlParse from 'git-url-parse';
+
+type Options = {
+  integrations: ScmIntegrationRegistry;
+};
 
 export class AnnotateLocationEntityProcessor implements CatalogProcessor {
+  constructor(private readonly options: Options) {}
+
   async preProcessEntity(
     entity: Entity,
     location: LocationSpec,
     _: CatalogProcessorEmit,
     originLocation: LocationSpec,
   ): Promise<Entity> {
-    return lodash.merge(
-      {
-        metadata: {
-          annotations: {
-            [LOCATION_ANNOTATION]: `${location.type}:${location.target}`,
-            [ORIGIN_LOCATION_ANNOTATION]: `${originLocation.type}:${originLocation.target}`,
-          },
-        },
-      },
-      entity,
-    );
+    const { integrations } = this.options;
+
+    const annotations: Record<string, string> = {
+      [LOCATION_ANNOTATION]: `${location.type}:${location.target}`,
+      [ORIGIN_LOCATION_ANNOTATION]: `${originLocation.type}:${originLocation.target}`,
+    };
+
+    if (isValidUrl(location.target)) {
+      const scmIntegration =
+        location.type === 'url'
+          ? integrations.byUrl(location.target)
+          : undefined;
+
+      annotations[VIEW_URL_ANNOTATION] = location.target;
+      const editUrl = scmIntegration?.resolveEditUrl(location.target);
+
+      if (editUrl) {
+        annotations[EDIT_URL_ANNOTATION] = editUrl;
+      }
+
+      annotations[SOURCE_LOCATION_ANNOTATION] = extractRepositoryRoot(
+        location.target,
+      );
+    }
+
+    return lodash.merge({ metadata: { annotations } }, entity);
   }
 }
+
+export function isValidUrl(value: string): boolean {
+  try {
+    // eslint-disable-next-line no-new
+    new URL(value);
+    return true;
+  } catch {
+    return false;
+  }
+}
+
+export function extractRepositoryRoot(url: string): string {
+  // TODO: Resolve URL from integrations doesn't actually support resolving to folders...
+  // In github one would have to switch from blob to tree urls...
+  // I wonder if this is something I should put into integrations.
+
+  const obj = gitUrlParse(url);
+
+  obj.filepath = '/';
+
+  return gitUrlParse.stringify(obj);
+}
```

Comment:
@freben I thought it would be a good idea if the "View Source" link navigates the user to the actual repository:

![image](https://user-images.githubusercontent.com/648527/109194723-0192d100-779a-11eb-9477-0c418a7b24a7.png)

I first thought that I can use resolveUrl from integration, but resolving to '/' doesn't actually work for GitHub has they use different URLs for viewing a folder. Alternative is to build an URL to the repository instead, or use this workaround. Not sure if that works for other providers. We could also move this to the integration package.

What do you think?

Reply:
These url dissection things are exactly the kind of stuff that fits well there!

Reply:
Would be cool to extend what we already have to support resolving `/` effectively, if that's useful 

https://github.com/backstage/backstage/blob/master/packages/integration/src/ScmIntegrations.ts#L86

Reply:
It would perfectly fit `resolveUrl`, but I don't think that it is possible. To generate a valid GitHub url I would have to know upfront whether I link to a folder or a file.

For example:

```
https://github.com/backstage/backstage/blob/master/packages/catalog-model/README.md
```

Can be resolved via `/` to: 

```
https://github.com/backstage/backstage/blob/master/
```

But that link is wrong, because it doesn't display the folder contents. It should be (note the `tree`):

```
https://github.com/backstage/backstage/tree/master/
```

But I don't think there is a way to decide whether a path is a file or a directory. What I could do, is to pass another argument that specifies that I expect the result to be a directory?

Reply:
Myeah I dunno. Using `tree` pointing to a file actually redirects to `blob` :P but yeah some of this functionality would actually require using the appropriate client lib instead of raw url manipulation to be safe

Reply:
GitLab has a redirect between both tree and blob which is nice. Bitbucket is even simpler and doesn't make a difference 🙂 

Not sure about azure, but neither could I implement the edit URL there because I miss access. Or do they have public repositories I can play around with?

Reply:
Hm not sure. But maybe just make it good enough for now (perhaps even just return the location unchanged) and let it evolve over time

Reply:
Good, that is the way I went 🙂 I think I also left enough comments there for the next one to finish the Azure part